waifu/datasets/soda.py

```diff
@@ -0,0 +1,36 @@
+import os
+import pickle
+import typing as t
+from dataclasses import dataclass
+
+import mashumaro
+import pandas as pd
+
+from waifu.datasets import BaseDataset
+from waifu.utils.dataset import get_data_path
+
+@dataclass(frozen=True)
+class SodaEpisode(mashumaro.DataClassDictMixin):
+    narrative: str
+    dialogue: t.List[str]
+    speakers: t.List[str]
```

Comment:
Something that I noticed: when the `relation` field is `xAttr`, the `literal` field usually contains a description about one of the speakers. Do you think we could incorporate both of those fields, and write a Persona Dialogue Module for rows where `relation == "xAttr"`, using `literal` as persona data in those cases?

Reply:
I guess you could? It's not like the `literal` datafield is very detailed, though. That might actually turn out in our favor, actually, because then the AI learns to also work with shitty one-liner personas decently enough. What do you think about that?

Reply:
I'm looking through the data again, actually. The literal in this case actually seems to describe the speaker acting as the _user_, not the description of the bot.

Reply:
> It's not like the literal datafield is very detailed, though. That might actually turn out in our favor, actually, because then the AI learns to also work with shitty one-liner personas decently enough. What do you think about that?

That's my intuition as well. Plus, the field always describes an attribute that is then immediately seen in the conversation, as opposed to the other datasets which have more detailed persona data but most of the conversation is only vaguely relevant to the persona. I think it'd be a nice addition to test out.

> The literal in this case actually seems to describe the speaker acting as the user, not the description of the bot.

The dataset is about dialogues between two people, so our pick of who's user vs. who's bot is pretty arbitrary, no? We could just swap the order around in the PDM I think.

Reply:
Sounds good, then. Would you want me to make a separate file for a SODA PDM like the others, or would you like me to convert the VDM file I already have into a PDM?

Reply:
I think converting into a PDM is fine - I don't think there's any point to repeating the SODA dataset but with persona data omitted (which is what a VDM would be) since the data augmentation already takes care of doing that for us.

Reply:
One more suggestion, as I'm coding the PDM. When I look at the data, there's more than a few conversations where one of the speakers isn't directly a person with a name, but things like "God", "Man with money", "Teacher" (among other examples pulled straight from the data). Under the current idea of swapping the order, this would mean that those users would be the abstract concepts. What I suggest is that we flip the speaker order _only_ when the relation is `xAttr`, and leave it the same everywhere else. Sounds good?

Reply:
Sounds fine to me.

Reply:
Actually, ignore that rambling. I'm a bit rarted.

Reply:
Implemented. Let me know if it looks good.

Reply:
LGTM! I'll merge as-is, and make the relevant refactoring changes related to the repo/package rename afterwards.